R/MsBackendDataFrame.R

```diff
@@ -95,80 +96,79 @@ setReplaceMethod("collisionEnergy", "MsBackendMemory", function(object, value) {
 })
 
 #' @rdname hidden_aliases
-setMethod("fromFile", "MsBackendMemory", function(object) {
+setMethod("fromFile", "MsBackendDataFrame", function(object) {
     if (any(colnames(object@spectraData) == "fromFile"))
         object@spectraData$fromFile
     else rep(NA_integer_, times = length(object))
 })
 
 #' @rdname hidden_aliases
-setMethod("intensity", "MsBackendMemory", function(object) {
+setMethod("intensity", "MsBackendDataFrame", function(object) {
     if (any(colnames(object@spectraData) == "intensity"))
         object@spectraData$intensity
     else {
-        lst <- list(numeric)
+        lst <- list(numeric())
         lst[rep(1, times = length(object))]
     }
 })
 
 #' @rdname hidden_aliases
-setMethod("ionCount", "MsBackendMemory", function(object) {
-    lapply(intensity(object), sum, na.rm = TRUE)
+setMethod("ionCount", "MsBackendDataFrame", function(object) {
+    vapply(intensity(object), sum, numeric(1), na.rm = TRUE)
 })
 
 #' @rdname hidden_aliases
-setMethod("isCentroided", "MsBackendMemory", function(object, ...) {
-    ## TODO @jo IMPLEMENT
-    stop("Not implemented for ", class(object), ".")
+setMethod("isCentroided", "MsBackendDataFrame", function(object, ...) {
+    vapply(peaks(object), .isCentroided, logical(1))
 })
 
 #' @rdname hidden_aliases
-setMethod("isEmpty", "MsBackendMemory", function(x) {
-    lengths(intensity(object)) == 0
+setMethod("isEmpty", "MsBackendDataFrame", function(x) {
+    lengths(intensity(x)) == 0
 })
 
 #' @rdname hidden_aliases
-setMethod("length", "MsBackendMemory", function(x) {
+setMethod("length", "MsBackendDataFrame", function(x) {
     nrow(x@spectraData)
 })
 
 #' @rdname hidden_aliases
-setMethod("msLevel", "MsBackendMemory", function(object, ...) {
+setMethod("msLevel", "MsBackendDataFrame", function(object, ...) {
     if (any(colnames(object@spectraData) == "msLevel"))
         object@spectraData$msLevel
     else rep(NA_integer_, times = length(object))
 })
 
 #' @rdname hidden_aliases
-setMethod("mz", "MsBackendMemory", function(object) {
+setMethod("mz", "MsBackendDataFrame", function(object) {
     if (any(colnames(object@spectraData) == "mz"))
         object@spectraData$mz
     else {
-        lst <- list(numeric)
+        lst <- list(numeric())
         lst[rep(1, times = length(object))]
     }
 })
 
 #' @rdname hidden_aliases
-setMethod("peaks", "MsBackendMemory", function(object) {
+setMethod("peaks", "MsBackendDataFrame", function(object) {
     mapply(mz(object), intensity(object), FUN = function(m, i)
-        data.frame(mz = mz, intensity = i))
+        data.frame(mz = m, intensity = i), SIMPLIFY = FALSE, USE.NAMES = FALSE)
```

Comment:
I would suggest that we return a `matrix` rather than a `data.frame` here. Any reason why you favour the latter?

Reply:
Yes. Acessing a column in a `data.frame` is faster than of a `matrix`. Also, creation *should* be faster and if I am not completely wrong `datafrane$mz` does not copy the content/vector while `matrix[, 1]` does... I believe.

Reply:
From the small test below, it looks like accessing a matrix column is faster, except if one uses `[[`.

```r
> data(itraqdata)
> dfr <- as(itraqdata[[1]], "data.frame")
> mat <- as.matrix(dfr)
microbenchmark(dfr[, 1], mat[, 1], dfr[, "mz"], mat[, "mz"], dfr[[1]], dfr[["mz"]])
Unit: microseconds
        expr   min      lq     mean  median      uq    max neval
    dfr[, 1] 9.779 10.3585 11.29819 10.8885 11.5125 45.016   100
    mat[, 1] 6.868  7.6930  8.23260  7.9920  8.6070 20.905   100
 dfr[, "mz"] 9.762 10.5505 11.18423 10.9155 11.7100 16.027   100
 mat[, "mz"] 6.803  7.6605  8.36979  8.0745  8.6530 33.461   100
    dfr[[1]] 4.964  5.6965  6.20307  6.0955  6.5190 10.753   100
 dfr[["mz"]] 5.038  5.7225  6.63447  6.1035  6.6800 39.812   100
```

Accessing a set of rows is considerably faster in matrices:

```r
> microbenchmark(dfr[20:99, ], mat[20:99, ], dfr[20:99, 1], mat[20:99, 1], dfr[[1]][20:99])
Unit: microseconds
            expr    min      lq     mean  median      uq     max neval
    dfr[20:99, ] 44.337 45.6870 47.01528 46.3155 46.9885  85.723   100
    mat[20:99, ]  2.069  2.5240  2.90010  2.8465  3.1115  10.600   100
   dfr[20:99, 1] 18.606 19.6800 21.57392 20.3080 20.9350 125.570   100
   mat[20:99, 1]  1.700  2.2745  2.61878  2.6245  2.9510   3.799   100
 dfr[[1]][20:99]  7.022  7.7185  8.64923  8.6995  9.0185  21.733   100
```
My initial suggestion to use a matrix rather than a data.frame was that we had only numerics. The above seems to suggest matrices would be better too. There might be other reasons to use dataframes.

What do you think?

Reply:
OK, then we go for `matrix`.